pubsub/google/cloud/pubsub_v1/publisher/batch/base.py

```diff
@@ -0,0 +1,145 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+
+import abc
+
+import six
+
+
+@six.add_metaclass(abc.ABCMeta)
+class Batch(object):
+    """The base batching class for Pub/Sub publishing.
+
+    Although the :class:`~.pubsub_v1.publisher.batch.thread.Batch` class, based
+    on :class:`threading.Thread`, is fine for most cases, advanced
+    users may need to implement something based on a different concurrency
+    model.
+
+    This class defines the interface for the Batch implementation;
+    subclasses may be passed as the ``batch_class`` argument to
+    :class:`~.pubsub_v1.client.PublisherClient`.
+
+    The batching behavior works like this: When the
+    :class:`~.pubsub_v1.publisher.client.Client` is asked to publish a new
+    message, it requires a batch. The client will see if there is an
+    already-opened batch for the given topic; if there is, then the message
+    is sent to that batch. If there is not, then a new batch is created
+    and the message put there.
+
+    When a new batch is created, it automatically starts a timer counting
+    down to the maximum latency before the batch should commit.
+    Essentially, if enough time passes, the batch automatically commits
+    regardless of how much is in it. However, if either the message count or
+    size thresholds are encountered first, then the batch will commit early.
+    """
+    def __len__(self):
+        """Return the number of messages currently in the batch."""
+        return len(self.messages)
+
+    @property
+    @abc.abstractmethod
+    def messages(self):
+        """Return the messages currently in the batch.
+
+        Returns:
+            Sequence: The messages currently in the batch.
+        """
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def size(self):
+        """Return the total size of all of the messages currently in the batch.
+
+        Returns:
+            int: The total size of all of the messages currently
+                 in the batch, in bytes.
+        """
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def settings(self):
+        """Return the batch settings.
+
+        Returns:
+            ~.pubsub_v1.types.BatchSettings: The batch settings. These are
+                considered immutable once the batch has been opened.
+        """
+        raise NotImplementedError
+
+    @property
```

Comment:
According to [the documentation](https://docs.python.org/3/library/abc.html#abc.abstractproperty), that form is deprecated and what I did is preferred (see the third paragraph at the link).